modules/sdk-core/src/bitgo/wallet/wallets.ts

```diff
@@ -538,6 +542,75 @@ export class Wallets implements IWallets {
       .result();
   }
 
+  /**
+   * Re-share wallet with existing spenders of the wallet
+   * @param walletId
+   * @param userPassword
+   */
+  async reshareOfcAccountWithSpenders(walletId: string, userPassword: string): Promise<void> {
+    const wallet = await this.bitgo.coin('ofc').wallets().get({ id: walletId });
+    if (!wallet?._wallet?.enterprise) {
+      throw new Error('Enterprise not found for the wallet');
+    }
+
+    const enterpriseUsersResponse = await this.bitgo
+      .get(`/api/v2/enterprise/${wallet?._wallet?.enterprise}/user`)
+      .result();
+    // create a map of users for easy lookup - we need the user email id to share the wallet
+    const usersMap = new Map(
+      [...enterpriseUsersResponse?.adminUsers, ...enterpriseUsersResponse?.nonAdminUsers].map((obj) => [obj.id, obj])
+    );
+    wallet?._wallet?.users?.forEach(async (user) => {
+      // user should be a spender and not an admin
+      if (user.permissions.includes('spend') && !user.permissions.includes('admin')) {
+        const userObject = usersMap.get(user.user);
+        const shareParams = {
+          walletId: walletId,
+          user: user.user,
+          permissions: user.permissions.join(','),
+          walletPassphrase: userPassword,
+          email: userObject.email.email,
+          coin: wallet.coin,
+          reshare: true,
+        };
+        wallet.shareWallet(shareParams);
+      }
+    });
+  }
+
+  /**
+   * Generate a random password
+   * @param   {Number} numWords     Number of 32-bit words
+   * @returns {String}          base58 random password
+   */
+  generateRandomPassword(numWords = 5): string {
+    const bytes = sjcl.codec.bytes.fromBits(sjcl.random.randomWords(numWords));
+    return bs58.encode(bytes);
+  }
+
+  /**
+   * Create keychain for ofc wallet using the password
+   * @param userPassword
+   * @returns
+   */
+  async createKeychain(userPassword: string): Promise<Keychain> {
+    const keychains = this.baseCoin.keychains();
+    const newKeychain = keychains.create();
+    const originalPasscodeEncryptionCode = this.generateRandomPassword();
+
+    const encryptedPrv = this.bitgo.encrypt({
+      password: userPassword,
+      input: newKeychain.prv,
+    });
+
+    const walletKeychain = await keychains.add({
```

Comment:
is the keychain also being added as part of the `updateShare` flow? @shivagupta94

Reply:
> is the keychain also being added as part of the updateShare flow? @shivagupta94
You mean in the wallet?

So, we need to create a keychain on the client side and save it in the backend using `keychains.add` method. Then, when triggering the updateShare method, we should pass the newly generated keychain ID in the payload. This way, we can replace it with the existing key ID.

Reply:
No. To the DB. This step - https://docs.google.com/document/d/1Dll1Z2OnVi2ofR__4LlqDdEI1aNUPuLXrqnITPQ0sf8/edit#bookmark=id.a9tdqvm0ql4f

Reply:
We're using the `keychains.add` function for this purpose. It saves the keychain details in the database and provides us with a reference ID for the same.

Reply:
> I was just asking if you're doing this during updateShare too and you're not

Yes we are adding keychain to DB on client side only

> if updateShare fails for any reason, then this key will be unused (not on any wallet)

I agree, implementing that approach would require setting up a rollback mechanism as well. Since creating a keychain is a prerequisite for this acceptshare flow, ensuring we can revert any erroneous actions would be essential.

Reply:
> I agree, implementing that approach would require setting up a rollback mechanism as well

You can use mongo transactions in WP for this